Add memory reporting infrastructure and use it to measure the display list.

nnethercote opened the pull request:
These changeset implements the beginnings of fine-grained measurement of Servo's data structures.

@pcwalton, @jdm: are you likely reviewers for this?

nnethercote committed:
Add memory reporting infrastructure and use it to measure the display list.

This changeset implements the beginnings of fine-grained measurement of
Servo's data structures.

- It adds a new `SizeOf` trait, which is used to measure the memory used
  by heap data structures, and implements it for some std types: Box,
  String, Option, Arc, Vec, and DList.

- It adds a new `MemoryReporter` trait which is used to report memory
  measurements from other threads to the memory profiler. Reporters are
  registered and unregistered with the memory profiler, and the memory
  profiler makes measurement requests of reporters when necessary.

- It plumbs a MemoryProfilerChan through to the layout task so it can
  register a memory reporter.

- It implements the `SizeOf` trait for `DisplayList` and associated
  types, and adds a memory reporter that uses it.

The display list hits 14.77 MiB when viewing
tests/html/perf-rainbow.html, and 2.51 MiB when viewing the Guardians of
the Galaxy Wikipedia page from servo-static-suite. Example output:

  0.29: display-list::http://www.reddit.com/
  0.00: display-list::http://static.adzerk.net/reddit/ads.html?sr=-reddit.com,loggedout&bust2#http://www.reddit.com
  0.00: display-list::http://www.reddit.com/static/createadframe.html

There are a number of FIXME comments indicating sub-optimal things. This
is a big enough change for now that doing them as follow-ups seems best.

nnethercote committed:
Rename LayoutTask::handle_script_request.

Because it handles requests from tasks other than the Script Task.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jdm commented:
I've left some comments on Critic.

jdm commented:
Note: I don't really know much about layout data strucutures, so someone else should definitely verify that all of the measurement there makes sense.

nnethercote committed:
Remove the fake memory profiler thread.

Currently if you don't specify the '-m' option, a fake do-nothing memory
profiler thread gets created instead of the real one. It ignores all
events except for `Exit`. And the timer thread doesn't get created so no
`Print` events are sent.

This changeset instead always creates the real thread. If you specify
'-m' the *timer* thread is still absent and it won't print anything.
However, the memory profiler thread will respond to register/unregister
events, which is good, because if there's a bug in those (e.g.
double-registration of a particular name) it'll show up in invocations
that lack '-m'.

nnethercote commented:
Who is the right person to review the layout parts? @pcwalton?

pcwalton commented:
I'll review the layout parts tomorrow.

jdm commented:
@pcwalton Tomorrow came and went.

pcwalton commented on components/gfx/display_list/mod.rs:
```diff
@@ -501,6 +513,14 @@ impl StackingContext {
+    fn size_of_excluding_self(&self) -> usize {
+        self.display_list.size_of_excluding_self()
+
+        // XXX: other fields may be measured later, esp. `layer`
```
nit: Servo's style is `FIXME(njn):` for this

pcwalton commented on components/gfx/display_list/mod.rs:
```diff
@@ -681,6 +708,18 @@ impl ClippingRegion {
+
+impl SizeOf for ComplexClippingRegion {
+    fn size_of_excluding_self(&self) -> usize {
+        0
```
add a FIXME?

nnethercote replied:
Not needed because ComplexClippingRegion doesn't have anything hanging off it (i.e. no Box fields).

pcwalton commented on components/gfx/display_list/mod.rs:
```diff
@@ -747,6 +798,13 @@ pub struct TextDisplayItem {
+impl SizeOf for TextDisplayItem {
+    fn size_of_excluding_self(&self) -> usize {
+        self.base.size_of_excluding_self()
+        // XXX: exclude `text_run` because it is non-owning?
```
nit: FIXME here too

pcwalton commented on components/gfx/display_list/mod.rs:
```diff
@@ -786,6 +851,20 @@ pub struct GradientDisplayItem {
+        // defined in this module, and we don't want to import GradientStop into util::memory where
+        // the SizeOf trait is defined. So we measure the elements directly.
+        self.base.size_of_excluding_self() +
+            heap_size_of(self.stops.as_ptr() as *const c_void)
```
One thing you can do for this kind of thing is to say `struct SizedGradientStops<'a>(&'a GradientStop); impl<'a> SizeOf for SizedGradientStops<'a> { ... }` to get around the coherence limitation. Then you can say `self.base.size_of_excluding_self() + SizedGradientStops(&self.stops).size_of_excluding_self()`. (This is a common Haskell trick; we call it "newtype wrapping".)

pcwalton commented on components/gfx/display_list/mod.rs:
```diff
@@ -1038,3 +1135,17 @@ impl fmt::Debug for DisplayItem {
+            BoxShadowClass(ref item)  => item.size_of_excluding_self(),
+        }
+    }
+}
```
Eventually with custom compiler macros it may well be nice to have this autogenerated, but obviously you don't have to do that right now.

pcwalton commented on components/layout/layout_task.rs:
```diff
@@ -436,6 +465,23 @@ impl LayoutTask {
+                                      &mut Option<MutexGuard<'a, LayoutTaskData>>) {
+        let mut reports = vec![];
+
+        // XXX: Just measuring the display tree for now.
```
`FIXME`

pcwalton commented on components/layout/layout_task.rs:
```diff
@@ -117,6 +119,9 @@ pub struct LayoutTask {
     pub id: PipelineId,
 
+    /// The URL of the pipeline that we belong to.
+    pub url: Url,
```
Isn't this in the `SharedLayoutContext`?

nnethercote replied:
As we discussed on IRC: there's no SharedLayoutContext readily available from the layout task's event loop, which is where it's needed. A SharedLayoutContext does get built during reflow, but it's built from data passed in via the `Reflow` message.

pcwalton commented:
Layout stuff reviewed, apologies for the delay.

nnethercote commented:
I've squashed the patches together and addressed all review comments. I think this is ready to merge, modulo test success.

jdm commented:
r+
